File: include/python/razel.py
# allows annotations with types that are not yet defined
from __future__ import annotations

import abc
import json
import os
from typing import ClassVar, Optional, Any, TypeVar
from collections.abc import Mapping, Sequence


class Razel:
    _instance: ClassVar[Optional[Razel]] = None

    def __init__(self, workspace_dir: str) -> None:
        self._workspace_dir = workspace_dir
        self._commands: list[Command] = []

    @staticmethod
    def init(workspace_dir: str) -> Razel:
        assert Razel._instance is None
        Razel._instance = Razel(workspace_dir)
        return Razel._instance

    @staticmethod
    def instance() -> Razel:
        assert Razel._instance is not None
        return Razel._instance

    def add_data_file(self, path: str) -> File:
        return File(self._rel_path(path), True, None)

    def add_output_file(self, path: str) -> File:
        return File(self._rel_path(path), False, None)

    def add_command(
        self, name: str, executable: str | File | Command, args: Sequence[str | File | Command],
        env: Optional[Mapping[str, str]] = None
    ) -> CustomCommand:
        name = self._sanitize_name(name)
        path = self._rel_path(_map_arg_to_output_path(executable))
        command = CustomCommand(name, path, _map_args_to_output_files(args), env)
        return self._add(command)

    def add_task(self, name: str, task: str, args: Sequence[str | File | Command]) -> Task:
        name = Razel._sanitize_name(name)
        command = Task(name, task, _map_args_to_output_files(args))
        return self._add(command)

    def ensure_equal(self, arg1: File | Command, arg2: File | Command) -> None:
        """Add a task to compare two files. In case of two commands, all output files will be compared."""
        if isinstance(arg1, Command) and isinstance(arg2, Command):
            assert len(arg1.outputs) == len(arg2.outputs)
            for i in range(len(arg1.outputs)):
                self.ensure_equal(arg1.outputs[i], arg2.outputs[i])
        else:
            file1 = _map_arg_to_output_file(arg1)
            file2 = _map_arg_to_output_file(arg2)
            name = f"{file1.basename}##shouldEqual##{file2.basename}"
            self._add(Task(name, "ensure-equal", [file1, file2]))

    def ensure_not_equal(self, arg1: File | Command, arg2: File | Command) -> None:
        """Add a task to compare two files. In case of two commands, all output files will be compared."""
        if isinstance(arg1, Command) and isinstance(arg2, Command):
            assert len(arg1.outputs) == len(arg2.outputs)
            for i in range(len(arg1.outputs)):
                self.ensure_equal(arg1.outputs[i], arg2.outputs[i])
        else:
            file1 = _map_arg_to_output_file(arg1)
            file2 = _map_arg_to_output_file(arg2)
            name = f"{file1.basename}##shouldNotEqual##{file2.basename}"
            self._add(Task(name, "ensure-not-equal", [file1, file2]))

    def write_razel_file(self) -> None:
        with open(os.path.join(self._workspace_dir, "razel.jsonl"), "w", encoding="utf-8") as file:
            for command in self._commands:
                json.dump(command.json(), file, separators=(',', ':'))
                file.write("\n")

    # Generic type used to ensure that _add() takes and returns the same type.
    _Command = TypeVar("_Command", bound="Command")

    def _add(self, command: _Command) -> _Command:
        for existing in self._commands:
            if existing.name == command.name:
                assert command.json() == existing.json(), \
                    f"conflicting actions: {command.name}:\nexisting: {existing.json()}\nto add: {command.json()}"
                assert isinstance(existing, type(command))
                return existing

        self._commands.append(command)
        return command

    @staticmethod
    def _sanitize_name(name: str) -> str:
        return name.replace(":", ".")  # target names may not contain ':'

    def _rel_path(self, file_name: str) -> str:
        if not os.path.isabs(file_name):
            return file_name

        return os.path.relpath(file_name, self._workspace_dir)


class File:
    def __init__(self, file_name: str, is_data: bool, created_by: Optional[Command]) -> None:
        self._file_name = file_name
        self._is_data = is_data
        self._created_by = created_by

    @property
    def file_name(self) -> str:
        return self._file_name

    @property
    def is_data(self) -> bool:
        return self._is_data

    @property
    def created_by(self) -> Optional[Command]:
        return self._created_by

    @property
    def basename(self) -> str:
        return os.path.basename(self._file_name)

    def ensure_equal(self, other: File | Command) -> None:
        Razel.instance().ensure_equal(self, other)

    def ensure_not_equal(self, other: File | Command) -> None:
        Razel.instance().ensure_not_equal(self, other)


class Command(abc.ABC):
    def __init__(self, name: str, inputs: Sequence[File], outputs: Sequence[File]) -> None:
        self._name = name
        self._inputs = inputs
        self._outputs = outputs
        self._stdout: File | None = None
        self._stderr: File | None = None
        for out in self._outputs:
            out._created_by = self

    @property
    def name(self) -> str:
        return self._name

    @property
    def inputs(self) -> Sequence[File]:
        return self._inputs

    @property
    def outputs(self) -> Sequence[File]:
        return self._outputs

    @property
    def output(self) -> File:
        assert len(self._outputs) == 1, \
            f"output() requires exactly one output file, but the command has {len(self._outputs)} outputs: {self.name}"
        return self._outputs[0]

    @property
    def stdout(self) -> File | None:
        return self._stdout

    @property
    def stderr(self) -> File | None:
        return self._stderr

    def ensure_equal(self, other: File | Command) -> None:
        Razel.instance().ensure_equal(self, other)

    def ensure_not_equal(self, other: File | Command) -> None:
        Razel.instance().ensure_not_equal(self, other)

    @abc.abstractmethod
    def json(self) -> Mapping[str, Any]:
        pass


class CustomCommand(Command):
    def __init__(
        self, name: str, executable: str, args: Sequence[str | File], env: Optional[Mapping[str, str]] = None
    ) -> None:
        (inputs, outputs) = _split_args_in_inputs_and_outputs(args)
        super().__init__(name, inputs, outputs)
        self._executable = executable
        self._args = args
        self._env = env

    @property
    def executable(self) -> str:
        return self._executable

    @property
    def args(self) -> Sequence[str | File]:
        return self._args

    @property
    def env(self) -> Optional[Mapping[str, str]]:
        return self._env

    def add_input_file(self, arg: str | File) -> CustomCommand:
        """Add an input file which is not part of the command line."""
        file = arg if isinstance(arg, File) else Razel.instance().add_data_file(arg)
        self._inputs.append(file)
        return self

    def add_input_files(self, args: Sequence[str | File]) -> CustomCommand:
        """Add input files which are not part of the command line."""
        for x in args:
            self.add_input_file(x)
        return self

    def add_output_file(self, arg: str | File) -> CustomCommand:
        """Add an output file which is not part of the command line."""
        file = arg if isinstance(arg, File) else Razel.instance().add_output_file(arg)
        file._createdBy = self
        self._outputs.append(file)
        return self

    def write_stdout_to_file(self, path: str = None) -> CustomCommand:
        self._stdout = Razel.instance().add_output_file(path if path else self._name)
        self._stdout._created_by = self
        self._outputs.append(self._stdout)
        return self

    def write_stderr_to_file(self, path: str = None) -> CustomCommand:
        self._stderr = Razel.instance().add_output_file(path if path else self._name)
        self._stderr._created_by = self
        self._outputs.append(self._stderr)
        return self

    def json(self) -> Mapping[str, Any]:
        j = {
            "name": self.name,
            "executable": self.executable,
            "args": [x.file_name if isinstance(x, File) else x for x in self.args],
            "inputs": [x.file_name for x in self._inputs],
            "outputs": [x.file_name for x in self._outputs if x != self._stdout and x != self._stderr],
        }
        if self.env:
            j["env"] = self.env
        if self._stdout:
            j["stdout"] = self._stdout.file_name
        if self._stderr:
            j["stderr"] = self._stderr.file_name
        return j


class Task(Command):
    @staticmethod
    def write_file(path: str, lines: Sequence[str]) -> File:
        file = Razel.instance().add_output_file(path)
        Razel.instance().add_task(path, "write-file", [file, *lines])
        return file

    def __init__(self, name: str, task: str, args: Sequence[str | File]) -> None:
        (inputs, outputs) = _split_args_in_inputs_and_outputs(args)
        super().__init__(name, inputs, outputs)
        self._task = task
        self._args = args

    @property
    def task(self) -> str:
        return self._task

    @property
    def args(self) -> Sequence[str | File]:
        return self._args

    def json(self) -> Mapping[str, Any]:
        return {
            "name": self.name,
            "task": self.task,
            "args": [x.file_name if isinstance(x, File) else x for x in self.args],
        }


def _map_arg_to_output_path(arg: str | File | Command) -> str:
    if isinstance(arg, Command):
        return arg.output.file_name
    elif isinstance(arg, File):
        return arg.file_name
    return arg


def _map_arg_to_output_file(arg: File | Command) -> File:
    return arg.output if isinstance(arg, Command) else arg


def _map_args_to_output_files(args: Sequence[str | File | Command]) -> Sequence[str | File]:
    return [x.output if isinstance(x, Command) else x for x in args]


def _split_args_in_inputs_and_outputs(args: Sequence[str | File | Command]) -> (Sequence[File], Sequence[File]):
    inputs = [x for x in args if isinstance(x, File) and (x.is_data or x.created_by)]
    outputs = [x for x in args if isinstance(x, File) and not x.is_data and x.created_by is None]
    return inputs, outputs

File: apis/deno/razel.ts
import { assert, assertEquals } from "https://deno.land/std@0.135.0/testing/asserts.ts";
import * as path from "https://deno.land/std@0.135.0/path/mod.ts";

export class Razel {
    static version = "0.5.6";
    private static _instance: Razel;
    razelFile: string;
    private commands: Command[] = [];

    private constructor(public readonly workspaceDir: string) {
        assert(path.isAbsolute(workspaceDir));
        this.razelFile = path.join(this.workspaceDir, "razel.jsonl");
    }

    static init(workspaceDir: string): Razel {
        assertEquals(Razel._instance, undefined);
        Razel._instance = new Razel(workspaceDir);
        return Razel._instance;
    }

    static instance(): Razel {
        return Razel._instance;
    }

    addDataFile(path: string): File {
        return new File(this.relPath(path), true, null);
    }

    addOutputFile(path: string): File {
        return new File(this.relPath(path), false, null);
    }

    addCommand(
        name: string,
        executable: string | File | Command,
        args: (string | File | Command)[],
        env?: any,
    ): CustomCommand {
        name = this.sanitizeName(name);
        const executablePath = mapArgToOutputPath(executable);
        const command = new CustomCommand(name, executablePath, mapArgsToOutputFiles(args), env || {});
        return this.add(command) as CustomCommand;
    }

    addTask(name: string, task: string, args: (string | File | Command)[]): Task {
        name = this.sanitizeName(name);
        const command = new Task(name, task, mapArgsToOutputFiles(args));
        return this.add(command) as Task;
    }

    // Add a task to compare two files. In case of two commands, all output files will be compared.
    ensureEqual(arg1: File | Command, arg2: File | Command): void {
        if (arg1 instanceof Command && arg2 instanceof Command) {
            assertEquals(
                arg1.outputs.length,
                arg2.outputs.length,
                "Commands to compare have different number of output files!",
            );
            for (let i = 0; i != arg1.outputs.length; ++i) {
                this.ensureEqual(arg1.outputs[i], arg2.outputs[i]);
            }
        } else {
            const file1 = mapArgToOutputFile(arg1);
            const file2 = mapArgToOutputFile(arg2);
            const name = `${file1.fileName}##shouldEqual##${file2.fileName}`;
            this.add(new Task(name, "ensure-equal", [file1, file2]));
        }
    }

    // Add a task to compare two files. In case of two commands, all output files will be compared.
    ensureNotEqual(arg1: File | Command, arg2: File | Command): void {
        if (arg1 instanceof Command && arg2 instanceof Command) {
            assertEquals(
                arg1.outputs.length,
                arg2.outputs.length,
                "Commands to compare have different number of output files!",
            );
            for (let i = 0; i != arg1.outputs.length; ++i) {
                this.ensureEqual(arg1.outputs[i], arg2.outputs[i]);
            }
        } else {
            const file1 = mapArgToOutputFile(arg1);
            const file2 = mapArgToOutputFile(arg2);
            const name = `${file1.fileName}##shouldNotEqual##${file2.fileName}`;
            this.add(new Task(name, "ensure-not-equal", [file1, file2]));
        }
    }

    /** Run the native razel binary to execute the commands.
     *
     * Commands are written to `<workspaceDir>/razel.jsonl`. That file is processed with `razel exec`.
     * If the native razel binary is not available, it will be downloaded.
     *
     * Output files are created in `<cwd>/razel-out`.
     */
    async run(args: string[] = ["exec"]) {
        await this.writeRazelFile();
        if (args.length > 0 && args[0] === "exec") {
            const razelFileRel = path.relative(Deno.cwd(), this.razelFile);
            args = [args[0], "-f", razelFileRel, ...args.slice(1)];
        }
        const status = await runRazelBinary(args);
        if (!status.success) {
            Deno.exit(status.code);
        }
    }

    async tryRun(args: string[] = ["exec"]): Promise<boolean> {
        await this.writeRazelFile();
        if (args.length > 0 && args[0] === "exec") {
            const razelFileRel = path.relative(Deno.cwd(), this.razelFile);
            args = [args[0], "-f", razelFileRel, ...args.slice(1)];
        }
        const status = await runRazelBinary(args);
        return status.success;
    }

    async writeRazelFile() {
        const json = this.commands.map((x) => JSON.stringify(x.json()));
        await Deno.writeTextFile(this.razelFile, json.join("\n") + "\n");
    }

    readLogFile(): LogFileItem[] {
        const path = "razel-out/razel-metadata/log.json";
        const text = Deno.readTextFileSync(path);
        return JSON.parse(text);
    }

    clear() {
        this.commands = [];
    }

    private add(command: Command): Command {
        const existing = this.commands.find((x) => x.name === command.name);
        if (existing) {
            const existingJson = existing.jsonForComparingToExistingCommand();
            const commandJson = command.jsonForComparingToExistingCommand();
            assertEquals(
                commandJson,
                existingJson,
                `conflicting command: ${command.name}:\n\
                existing: ${JSON.stringify(existingJson)}\n\
                to add:   ${JSON.stringify(commandJson)}`,
            );
            return existing;
        }
        this.commands.push(command);
        return command;
    }

    private sanitizeName(name: string): string {
        return name.replaceAll(":", "."); // target names may not contain ':'
    }

    private relPath(fileName: string): string {
        if (!path.isAbsolute(fileName) || !fileName.startsWith(this.workspaceDir)) {
            return fileName;
        }
        return path.relative(this.workspaceDir, fileName);
    }
}

export namespace Razel {
    export enum Tag {
        // don't be verbose if command succeeded
        Quiet = "razel:quiet",
        // always show verbose output
        Verbose = "razel:verbose",
        // keep running and don't be verbose if command failed
        Condition = "razel:condition",
        // always execute a command without caching
        NoCache = "razel:no-cache",
        // don't use remote cache
        NoRemoteCache = "razel:no-remote-cache",
        // disable sandbox and also cache - for commands with unspecified input/output files
        NoSandbox = "razel:no-sandbox",
    }
}

export class File {
    constructor(
        public readonly fileName: string,
        public readonly isData: boolean,
        public createdBy: Command | null,
    ) {
    }

    get basename(): string {
        return path.basename(this.fileName);
    }

    ensureEqual(other: File | Command): void {
        Razel.instance().ensureEqual(this, other);
    }

    ensureNotEqual(other: File | Command): void {
        Razel.instance().ensureNotEqual(this, other);
    }
}

export abstract class Command {
    public stdout: File | undefined = undefined;
    public stderr: File | undefined = undefined;
    public readonly deps: Command[] = [];
    public readonly tags: (Razel.Tag | string)[] = [];

    protected constructor(
        public readonly name: string,
        public readonly inputs: File[],
        public readonly outputs: File[],
    ) {
        this.outputs.forEach((x) => x.createdBy = this);
    }

    get output(): File {
        assertEquals(
            this.outputs.length,
            1,
            `output() requires exactly one output file, but the command has ${this.outputs.length} outputs: ${this.name}`,
        );
        return this.outputs[0];
    }

    addDependency(dependency: Command): this {
        if (!this.deps.includes(dependency)) {
            this.deps.push(dependency);
        }
        return this;
    }

    addDependencies(dependencies: Command[]): this {
        dependencies.forEach((x) => this.addDependency(x));
        return this;
    }

    addTag(tag: Razel.Tag | string): this {
        if (!this.tags.includes(tag)) {
            this.tags.push(tag);
        }
        return this;
    }

    addTags(tags: (Razel.Tag | string)[]): this {
        tags.forEach((x) => this.addTag(x));
        return this;
    }

    ensureEqual(other: File | Command): void {
        Razel.instance().ensureEqual(this, other);
    }

    ensureNotEqual(other: File | Command): void {
        Razel.instance().ensureNotEqual(this, other);
    }

    abstract json(): any;

    abstract jsonForComparingToExistingCommand(): any;
}

export class CustomCommand extends Command {
    constructor(
        name: string,
        public readonly executable: string,
        public readonly args: (string | File)[],
        public readonly env: any,
    ) {
        const [inputs, outputs] = splitArgsInInputsAndOutputs(args);
        super(name, inputs, outputs);
    }

    addEnv(key: string, value: string): CustomCommand {
        this.env[key] = value;
        return this;
    }

    // Add an input file which is not part of the command line.
    addInputFile(arg: string | File): CustomCommand {
        const file = arg instanceof File ? arg : Razel.instance().addDataFile(arg);
        if (!this.inputs.some((x) => x.fileName === file.fileName)) {
            this.inputs.push(file);
        }
        return this;
    }

    // Add input files which are not part of the command line.
    addInputFiles(args: (string | File)[]): CustomCommand {
        args.forEach((x) => this.addInputFile(x));
        return this;
    }

    // Add an output file which is not part of the command line.
    addOutputFile(arg: string | File): CustomCommand {
        const file = arg instanceof File ? arg : Razel.instance().addOutputFile(arg);
        if (!this.outputs.some((x) => x.fileName === file.fileName)) {
            file.createdBy = this;
            this.outputs.push(file);
        }
        return this;
    }

    writeStdoutToFile(path?: string): CustomCommand {
        const newFile = Razel.instance().addOutputFile(path ? path : this.name + ".stdout.txt");
        if (this.stdout) {
            assertEquals(newFile.fileName, this.stdout.fileName);
            return this;
        }
        this.stdout = newFile;
        this.stdout.createdBy = this;
        this.outputs.push(this.stdout);
        return this;
    }

    writeStderrToFile(path?: string): CustomCommand {
        const newFile = Razel.instance().addOutputFile(path ? path : this.name + ".stderr.txt");
        if (this.stderr) {
            assertEquals(newFile.fileName, this.stderr.fileName);
            return this;
        }
        this.stderr = newFile;
        this.stderr.createdBy = this;
        this.outputs.push(this.stderr);
        return this;
    }

    json(): any {
        return {
            name: this.name,
            executable: this.executable,
            args: this.args.map((x) => x instanceof File ? x.fileName : x),
            inputs: this.inputs.map((x) => x.fileName),
            outputs: this.outputs.filter((x) => x !== this.stdout && x !== this.stderr).map((x) =>
                x.fileName
            ),
            env: Object.keys(this.env).length !== 0 ? this.env : undefined,
            stdout: this.stdout?.fileName,
            stderr: this.stderr?.fileName,
            deps: this.deps.length != 0 ? this.deps.map((x) => x.name) : undefined,
            tags: this.tags.length != 0 ? this.tags : undefined,
        };
    }

    jsonForComparingToExistingCommand(): any {
        return {
            executable: this.executable,
            args: this.args.map((x) => x instanceof File ? x.fileName : x),
            // additional input/output files might be added after constructor(), therefore not adding them here
            // additional env variables might be added after constructor(), therefore not adding them here
        };
    }
}

export class Task extends Command {
    static writeFile(path: string, lines: string[]): File {
        const file = Razel.instance().addOutputFile(path);
        Razel.instance().addTask(path, "write-file", [file, ...lines]);
        return file;
    }

    constructor(name: string, public readonly task: string, public readonly args: (string | File)[]) {
        const [inputs, outputs] = splitArgsInInputsAndOutputs(args);
        super(name, inputs, outputs);
    }

    json(): any {
        return {
            name: this.name,
            task: this.task,
            args: this.args.map((x) => x instanceof File ? x.fileName : x),
            deps: this.deps.length != 0 ? this.deps.map((x) => x.name) : undefined,
            tags: this.tags.length != 0 ? this.tags : undefined,
        };
    }

    jsonForComparingToExistingCommand(): any {
        return {
            task: this.task,
            args: this.args.map((x) => x instanceof File ? x.fileName : x),
        };
    }
}

export interface LogFileItem {
    name: string;
    tags?: string[];
    status: string;
    error?: string;
    cache?: string;
    // original execution duration of the command/task - ignoring cache
    exec?: number;
    // actual duration of processing the command/task - including caching and overheads
    total?: number;
    // total size of all output files and stdout/stderr [bytes]
    output_size?: number;
    measurements?: { [key: string]: string | number };
}

function mapArgToOutputPath(arg: string | File | Command): string {
    if (arg instanceof Command) {
        return arg.output.fileName;
    } else if (arg instanceof File) {
        return arg.fileName;
    }
    return arg;
}

function mapArgToOutputFile(arg: File | Command): File {
    return arg instanceof Command ? arg.output : arg;
}

function mapArgsToOutputFiles(args: (string | File | Command)[]): (string | File)[] {
    return args.map((x) => x instanceof Command ? x.output : x);
}

function splitArgsInInputsAndOutputs(args: (string | File)[]): [File[], File[]] {
    const inputs = args.filter((x) =>
        (x instanceof File) && ((x as File).isData || (x as File).createdBy)
    ) as File[];
    const outputs = args.filter((x) =>
        (x instanceof File) && !(x as File).isData && !(x as File).createdBy
    ) as File[];
    return [inputs, outputs];
}

export async function findOrDownloadRazelBinary(version: string): Promise<string> {
    const ext = Deno.build.os === "windows" ? ".exe" : "";
    // try to use razel binary from PATH
    let razelBinaryPath = `razel${ext}`;
    if (await getRazelVersion(razelBinaryPath) === version) {
        return razelBinaryPath;
    }
    // try to use razel binary from .cache
    let cacheDir;
    if (Deno.build.os === "darwin") {
        cacheDir = `${Deno.env.get("HOME")}/Library/Caches/de.reu-dev.razel`;
    } else if (Deno.build.os === "windows") {
        const localAppData = Deno.env.get("LOCALAPPDATA");
        assert(localAppData);
        cacheDir = `${localAppData.replaceAll("\\", "/")}/reu-dev/razel`;
    } else {
        cacheDir = `${Deno.env.get("HOME")}/.cache/razel`;
    }
    razelBinaryPath = `${cacheDir}/razel${ext}`;
    if (await getRazelVersion(razelBinaryPath) === version) {
        return razelBinaryPath;
    }
    // download razel binary to .cache
    await downloadRazelBinary(version, razelBinaryPath);
    return razelBinaryPath;
}

async function getRazelVersion(razelBinaryPath: string): Promise<string | null> {
    try {
        const command = new Deno.Command(razelBinaryPath, {
            args: ["--version"],
            stdout: "piped",
            stderr: "piped",
        });
        const { code, stdout } = await command.output();
        if (code !== 0) {
            return null;
        }
        const stdoutDecoded = new TextDecoder().decode(stdout);
        return stdoutDecoded.trim().split(" ")[1];
    } catch {
        return null;
    }
}

async function downloadRazelBinary(version: string | null, razelBinaryPath: string) {
    const downloadTag = version ? `download/v${version}` : "latest/download";
    let buildTarget;
    if (Deno.build.os === "darwin" && Deno.build.arch === "aarch64") {
        buildTarget = "aarch64-apple-darwin";
    } else if (Deno.build.os === "darwin" && Deno.build.arch === "x86_64") {
        buildTarget = "x86_64-apple-darwin";
    } else if (Deno.build.os === "windows") {
        buildTarget = "x86_64-pc-windows-msvc";
    } else if (Deno.build.arch === "aarch64") {
        buildTarget = "aarch64-unknown-linux-gnu";
    } else {
        buildTarget = "x86_64-unknown-linux-gnu";
    }
    const url = `https://github.com/reu-dev/razel/releases/${downloadTag}/razel-${buildTarget}.gz`;
    console.log("Download razel binary from", url);
    const response = await fetch(url);
    if (!response.body) {
        throw response.statusText;
    }
    console.log(`Extract razel binary to ${razelBinaryPath}`);
    await Deno.mkdir(path.dirname(razelBinaryPath), { recursive: true });
    const dest = await Deno.open(razelBinaryPath, { create: true, write: true });
    await response.body
        .pipeThrough(new DecompressionStream("gzip"))
        .pipeTo(dest.writable);
    if (Deno.build.os !== "windows") {
        const mode = (await Deno.stat(razelBinaryPath)).mode || 0;
        await Deno.chmod(razelBinaryPath, mode | 0o700);
    }
    const actualVersion = await getRazelVersion(razelBinaryPath);
    assert(
        actualVersion,
        "Failed to download razel binary. To build it from source, run: cargo install razel",
    );
    console.log(`Downloaded razel ${actualVersion}`);
}

// Run the native razel binary. If not available, it will be downloaded. Returns the exit code.
export async function runRazelBinary(args: string[]): Promise<Deno.CommandStatus> {
    const razelBinaryPath = await findOrDownloadRazelBinary(Razel.version);
    const cmd = [razelBinaryPath, ...args];
    console.log(cmd.join(" "));
    const process = new Deno.Command(cmd[0], {
        args: cmd.slice(1),
        stdout: "inherit",
        stderr: "inherit",
    }).spawn();
    return await process.status;
}

if (import.meta.main) {
    if (Deno.args.length === 0) {
        await findOrDownloadRazelBinary(Razel.version);
        Deno.exit(0);
    }
    const status = await runRazelBinary(Deno.args);
    Deno.exit(status.code);
}
